### PR TITLE
Pin `Sphinx<4.0`

### LINKED ISSUE #7455: Docs build error in Sphinx 4.0.0
## 🐛 Bug
I locally confirmed that building docs (`cd docs; make html`) fails due to:
```
Extension error (sphinx_paramlinks.sphinx_paramlinks):
Handler <function build_index at 0x7f20c201c820> for event 'doctree-read' threw an exception (exception: __new__() missing 1 required positional argument: 'aliased')
make: *** [Makefile:19: html] Error 2
Traceback (most recent call last):
  File "/entrypoint.py", line 22, in <module>
    action.build_all_docs(github_env, [os.environ.get("INPUT_DOCS-FOLDER")])
  File "/sphinx_action/action.py", line 167, in build_all_docs
    raise RuntimeError("Build failed")
RuntimeError: Build failed
```
and also confirmed that `pip install Sphinx<4.0.0` fixes the error.

### To Reproduce
https://github.com/PyTorchLightning/pytorch-lightning/pull/7452/checks?check_run_id=2538915517

### Expected behavior
`make html` should successfully build the docs.

### PR DESCRIPTION
**Pull request by akihironitta**
## What does this PR do?
Fixes #7455. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by akihironitta** [comment]
I'm now ooking into why the following error happens...
```
Running Sphinx v2.4.4

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/sphinx/config.py", line 348, in eval_config_file
    execfile_(filename, namespace)
  File "/usr/local/lib/python3.8/site-packages/sphinx/util/pycompat.py", line 81, in execfile_
    exec(code, _globals)
  File "/github/workspace/docs/source/conf.py", line 22, in <module>
    import pt_lightning_sphinx_theme
ModuleNotFoundError: No module named 'pt_lightning_sphinx_theme'
```

https://github.com/PyTorchLightning/pytorch-lightning/pull/7456/checks?check_run_id=2539437267

**Review by Borda — APPROVED** [review]

File: requirements/docs.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-sphinx>=3.0, !=3.5  # fails with sphinx.ext.viewcode
+sphinx>=3.0, !=3.5, <4.0  # fails with sphinx.ext.viewcode  # fails with sphinx_paramlinks with 4.0.0
```
Comment (Borda) [inline requirements/docs.txt]: There was no other release between 3.5 and 4.0

Comment (akihironitta) [reply requirements/docs.txt]: I was wondering why it was not using `!=3.5` in the very last successful runs even though we specified `!=3.5`:
```
Sphinx                        3.5.4
```
https://github.com/PyTorchLightning/pytorch-lightning/runs/2535513652

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by awaelchli — APPROVED** [review]
need to wait for a sphinx-paramlinks update like last time  :)

**Pull request merged**